File: src/backend/modules/m4_finance/schemas/bank_import.py
"""
M4 재무/회계 — 은행 입금 내역 임포트 Pydantic 스키마
"""
from datetime import date
from typing import Optional
from pydantic import BaseModel, Field


# ── CSV 파싱 결과 (미리보기용) ──

class ParsedTransaction(BaseModel):
    """CSV에서 파싱된 개별 거래"""
    row_index: int = Field(..., description="원본 CSV 행 번호")
    transaction_date: str = Field(..., description="거래일자 (YYYY-MM-DD)")
    description: str = Field("", description="적요")
    deposit_amount: float = Field(0, description="입금액")
    withdrawal_amount: float = Field(0, description="출금액")
    balance: float = Field(0, description="거래후 잔액")
    hash: str = Field(..., description="중복 감지용 해시")
    is_duplicate: bool = Field(False, description="기존 임포트 중복 여부")
    mapped_account_id: Optional[str] = Field(None, description="자동 매핑된 계정과목 ID")
    mapped_account_name: Optional[str] = Field(None, description="자동 매핑된 계정과목명")


class ParseResult(BaseModel):
    """CSV 파싱 전체 결과"""
    bank_code: str
    file_name: str
    total_rows: int = Field(0, description="전체 행 수")
    deposit_rows: int = Field(0, description="입금 건 수")
    transactions: list[ParsedTransaction] = []


# ── 전표 생성 확인 요청 ──

class ConfirmTransaction(BaseModel):
    """사용자가 선택/수정한 개별 거래"""
    transaction_date: str = Field(..., description="거래일자 (YYYY-MM-DD)")
    description: str = Field("", description="적요")
    amount: float = Field(..., description="입금액")
    account_id: str = Field(..., description="계정과목 ID (대변)")
    hash: str = Field(..., description="중복 방지용 해시")


class ConfirmImportRequest(BaseModel):
    """전표 생성 확인 요청"""
    bank_code: str = Field(..., description="은행 코드")
    file_name: str = Field("", description="원본 파일명")
    bank_account_id: Optional[str] = Field(None, description="보통예금 계정과목 ID (차변)")
    transactions: list[ConfirmTransaction] = Field(..., min_length=1)


# ── 매핑 규칙 ──

class MappingCreate(BaseModel):
    """적요→계정과목 매핑 규칙 생성"""
    keyword: str = Field(..., min_length=1, description="적요 키워드")
    account_id: str = Field(..., description="계정과목 ID")
    priority: int = Field(0, description="우선순위 (높을수록 우선)")


class MappingResponse(BaseModel):
    """매핑 규칙 응답"""
    id: str
    keyword: str
    account_id: str
    account_name: Optional[str] = None
    priority: int = 0
    is_active: bool = True

File: src/frontend/src/api/finance/bankImport.ts
/**
 * M4 재무/회계 — 은행 입금 내역 임포트 API
 */
import api from '../client';

/* ── 타입 정의 ── */

export interface ParsedTransaction {
  row_index: number;
  transaction_date: string;
  description: string;
  deposit_amount: number;
  withdrawal_amount: number;
  balance: number;
  hash: string;
  is_duplicate: boolean;
  mapped_account_id: string | null;
  mapped_account_name: string | null;
}

export interface ParseResult {
  bank_code: string;
  file_name: string;
  total_rows: number;
  deposit_rows: number;
  transactions: ParsedTransaction[];
}

export interface ConfirmTransaction {
  transaction_date: string;
  description: string;
  amount: number;
  account_id: string;
  hash: string;
}

export interface ImportHistory {
  id: string;
  import_date: string;
  bank_code: string;
  file_name: string;
  total_rows: number;
  imported_count: number;
  skipped_count: number;
  total_deposit: number;
  source: string;
}

export interface AccountMapping {
  id: string;
  keyword: string;
  account_id: string;
  account_name: string | null;
  priority: number;
  is_active: boolean;
}

export interface CompanyBankAccount {
  id: string;
  bank_code: string;
  bank_name: string;
  account_no: string;
  account_holder: string;
  account_type: string;
  chart_account_id: string;
  chart_account_name: string | null;
  is_primary: boolean;
  is_active: boolean;
  memo: string | null;
}

/* ── API 함수 ── */

/** CSV 파일 업로드 + 파싱 (미리보기) */
export async function parseBankCSV(file: File, bankCode: string): Promise<ParseResult> {
  const formData = new FormData();
  formData.append('file', file);
  const res = await api.post(`/finance/bank-import/parse?bank_code=${bankCode}`, formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  });
  return res.data?.data ?? res.data;
}

/** 전표 생성 확인 */
export async function confirmImport(data: {
  bank_code: string;
  file_name: string;
  bank_account_id?: string;
  transactions: ConfirmTransaction[];
}) {
  const res = await api.post('/finance/bank-import/confirm', data);
  return res.data?.data ?? res.data;
}

/** 임포트 이력 조회 */
export async function getImportHistory(params?: { page?: number; size?: number }) {
  const res = await api.get('/finance/bank-import/history', { params });
  return res.data?.data ?? res.data;
}

/** 매핑 규칙 목록 */
export async function getMappings(): Promise<AccountMapping[]> {
  const res = await api.get('/finance/bank-import/mappings');
  return res.data?.data ?? res.data;
}

/** 매핑 규칙 추가 */
export async function createMapping(data: {
  keyword: string;
  account_id: string;
  priority?: number;
}): Promise<AccountMapping> {
  const res = await api.post('/finance/bank-import/mappings', data);
  return res.data?.data ?? res.data;
}

/** 매핑 규칙 삭제 */
export async function deleteMapping(id: string) {
  const res = await api.delete(`/finance/bank-import/mappings/${id}`);
  return res.data?.data ?? res.data;
}

/* ── 회사 은행 계좌 관리 ── */

/** 등록된 회사 계좌 목록 */
export async function getBankAccounts(): Promise<CompanyBankAccount[]> {
  const res = await api.get('/finance/bank-import/accounts');
  return res.data?.data ?? res.data;
}

/** 회사 계좌 등록 */
export async function createBankAccount(data: {
  bank_code: string;
  bank_name: string;
  account_no: string;
  account_holder: string;
  account_type?: string;
  chart_account_id: string;
  is_primary?: boolean;
  memo?: string;
}): Promise<CompanyBankAccount> {
  const res = await api.post('/finance/bank-import/accounts', data);
  return res.data?.data ?? res.data;
}

/** 회사 계좌 수정 */
export async function updateBankAccount(id: string, data: Partial<{
  bank_code: string;
  bank_name: string;
  account_no: string;
  account_holder: string;
  account_type: string;
  chart_account_id: string;
  is_primary: boolean;
  memo: string;
}>): Promise<CompanyBankAccount> {
  const res = await api.put(`/finance/bank-import/accounts/${id}`, data);
  return res.data?.data ?? res.data;
}

/** 회사 계좌 삭제 */
export async function deleteBankAccount(id: string) {
  const res = await api.delete(`/finance/bank-import/accounts/${id}`);
  return res.data?.data ?? res.data;
}
